app/src/main/java/org/mozilla/tv/firefox/webrender/cursor/CursorView.kt

```diff
@@ -5,89 +5,103 @@
+
+        Observables.combineLatest(cursorModel.isCursorMoving, cursorModel.isSelectPressed) {
+            // Only emit false if we are both stationary and not pressed
+            moving, pressed -> moving || pressed
         }
-    }
+                .distinctUntilChanged()
+                .subscribe { cursorIsMovingOrPressed ->
+                    if (cursorIsMovingOrPressed) {
+                        animate().cancel()
+                        alpha = 1f
+                    } else {
+                        animate()
+                                .setStartDelay(HIDE_AFTER_MILLIS)
+                                .setDuration(HIDE_ANIMATION_DURATION_MILLIS)
+                                .alpha(0f)
+                                .start()
+                    }
+                }.addTo(compositeDisposable)
 
-    fun cancelUpdates() {
-        animate().cancel()
-        hideHandler.removeMessages(HIDE_MESSAGE_ID)
-    }
+        cursorModel.isSelectPressed
+                .distinctUntilChanged()
+                .subscribe { pressed ->
+                    when (pressed) {
+                        true -> setImageResource(R.drawable.cursor_full_active)
+                        false -> setImageResource(R.drawable.cursor_full)
+                    }
+                }.addTo(compositeDisposable)
 
-    fun startUpdates() {
-        setMaxVisibility()
-        resetCountdown()
+        return compositeDisposable
     }
 
-    private fun setMaxVisibility() {
-        animate().cancel()
-        alpha = 1f
+    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
+        super.onSizeChanged(w, h, oldw, oldh)
+        this.width = w.toFloat()
+        this.height = h.toFloat()
     }
 
-    private fun resetCountdown() {
-        hideHandler.removeMessages(HIDE_MESSAGE_ID)
-        hideHandler.sendEmptyMessageDelayed(HIDE_MESSAGE_ID, HIDE_AFTER_MILLIS)
+    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
+        super.onLayout(changed, left, top, right, bottom)
+        position.x = x
+        position.y = y
     }
-}
 
-/**
- * Hides the cursor when it receives a message.
- *
- * We use a [Handler], with [Message]s, because they make no allocations, unlike
- * more modern/readable approaches:
- * - coroutines
- * - Animators with start delays (and cancelling them as necessary)
- */
-private class CursorHideHandler(view: CursorView) : Handler(Looper.getMainLooper()) {
-    private val viewWeakReference = WeakReference<CursorView>(view)
-
-    override fun handleMessage(msg: Message?) {
-        viewWeakReference.get()
-                ?.animate()
-                ?.setDuration(HIDE_ANIMATION_DURATION_MILLIS)
-                ?.alpha(0f)
-                ?.start()
+    override fun onDraw(canvas: Canvas?) {
+        super.onDraw(canvas)
+
+        val shouldInvalidate = cursorModel?.mutatePosition(position) ?: true
+        x = position.x - (width / 2)
```

Comment:
We do need to set `View.x` and `View.y` at some point.  I suppose we could pass the View itself into the model, instead of a `PointF`.

Reply:
My solution would do that under "...":
```kotlin
val onDrawMutablePositionCache = PointF()

override fun onDraw(canvas: Canvas?) {
    onDrawMutablePositionCache.set(x, y)
    val shouldInvalidate = cursorModel?.mutatePosition(onDrawMutablePositionCache) ?: false
    x = onDrawMutablePositionCache.x
    y = onDrawableMutablePositionCache.y
    // ...
```

Reply:
This could me being dense, so apologies if I'm missing the point here, but I'm not clear on what this solves.  As it is, `position` (which seems to be equivalent to `onDrawMutablePositionCache`) is our source of truth.  We mutate it and read it to see where we should draw.  `View.x` and `View.y` are only read in `onLayout` (and looking more closely, that can be removed entirely).  `position` is our source of truth, `x` and `y` are set only for their side effects.

Again, possible that I'm just missing your point.

Reply:
They serve a similar purpose: both cache a position in order to avoid allocation. However, the difference is about scoping: my suggestion is scope to the `onDraw` function (which is sadly not enforced by the compiler) while `position` scopes to the entire class.

My concern is that by scoping to the entire class, we're forced to duplicate a real source of truth: `View.x/y`. While I'm not always against doing that, it creates the burden of having to accurately duplicate state: you have to find all the places those values can change and override them (e.g. your code in `onLayout` would be unnecessary in my solution). If you miss a spot, you end up with bugs (e.g. a very real problem in the current implementation is `CursorModel` or `WebRenderFragment` could call `setX/Y` and break the implementation).

Reply:
That's a good point, I'll make the change.